src/dmd/backend/outbuf.d

```diff
@@ -90,34 +89,45 @@ struct Outbuffer
 
     // Reserve nbytes in buffer
     @trusted
-    void enlarge(size_t nbytes)
+    private void enlarge(size_t nbytes)
     {
         pragma(inline, false);  // do not inline slow path
-        const size_t oldlen = pend - buf;
-        const size_t used = p - buf;
-
-        size_t len = used + nbytes;
-        // No need to reallocate
-        if (nbytes < (pend - p))
-            return;
 
-        const size_t newlen = oldlen + (oldlen >> 1);   // oldlen * 1.5
-        if (len < newlen)
-            len = newlen;
-        len = (len + 15) & ~15;
+        debug assert(nbytes > pend - p, "You should call this function from reserve() only.");
```

Comment:
so we have a tiny function kept tiny for inlining, which then calls a large function which should only be called from said tiny function which enforces that by checking you didn't ask it for too little?

Idk, man...

If the intention is to not trash icache on calling `reserve` when there is spare capacity (presumably the common case), then `pragma(inline, false)` should go on `enlarge`. Could even make it a local function to `enlarge` as well 😱.

Reply:
> then pragma(inline, false) should go on enlarge.

`enlarge` is already noinline.
```
    private void enlarge(size_t nbytes)
    {
        pragma(inline, false);  // do not inline slow path
```

Making it a local function does make more sense than this assert though.

Reply:
Yah, that's sort of a classic idiom. The idea is that `reserve()` is called from all over the place and has a small frequent case and on occasion defers to the non-inline call.

I thought of actually defining `enlarge` as a static function inside `reserve` but that would be a bit too cute.

Reply:
Very much aware of the pattern. I just completely missed the pragma.

I think it being inside `reserve` makes perfect sense but whatever, it's not important.

Reply:
@John-Colvin that's all the encouragement I needed. What does it look like now?

Reply:
Seems reasonable, but the testsuite still says no.